File: swift/Sources/FantasticIoBridge/Authorizer.swift
// Bridge authorization — base types + the two rule registries (ingress/egress).
//
// Swift mirror of the canonical Python `bridge_core` (ingress_rules / egress_rules
// packages) and the Rust `authorizer` module. Two independent, TYPED rules govern a
// leg, mirrored on the wire and enforced on the RECEIVER:
//   - an INGRESS rule (the inbound FILTER): `authorize(action) -> AuthDecision`.
//   - an EGRESS rule (the outbound DECORATOR): `credential() -> token?`, stamped on
//     the frame ENVELOPE by `forward` (never the dispatched payload).
//
// Each rule is typed in the record (`{"type": <name>, "env": <var>}`) and resolved
// BY NAME from a registry — the `IngressRules` / `EgressRules` namespaces (a folder
// of one-rule-per-file, the `resolve` switch is the importer). The record carries
// `ingress_rule`/`egress_rule` (symmetric) or the legacy `auth` shorthand (both).
//
// SWIFT SPECIFIC: there is no shared read loop. The ONLY inbound-`call` dispatcher is
// `RelayTransport.dispatch`, so an ingress rule is enforced on the relay_connector
// leg only (correct-by-construction; ws is an asymmetric client, in-memory forward is
// a direct kernel call). The relay e2e matrix covers Swift's real inbound path.
// Rules are TRANSITIONAL (inline plumbing), not invocational (agents).

import FantasticJSON
import Foundation

/// Default env var for the `password` rule when the record names none.
public let DefaultGroupTokenEnv = "FANTASTIC_GROUP_TOKEN"

/// One inbound request the peer is asking this leg to perform locally.
public struct AuthAction: Sendable {
    /// `"call"` (gated) | `"watch"` | `"unwatch"`.
    public let kind: String
    /// The local agent id the peer addressed.
    public let target: String
    /// `payload["type"]` — the verb requested (e.g. `"reflect"`).
    public let verb: String
    /// The `auth_token` the peer attached on the frame ENVELOPE, if any (read by the
    /// `password` rule; the dispatched payload never carries it).
    public let token: String?

    public init(kind: String, target: String, verb: String, token: String? = nil) {
        self.kind = kind
        self.target = target
        self.verb = verb
        self.token = token
    }
}

/// An ingress rule's verdict on an `AuthAction`.
public enum AuthDecision: Sendable {
    /// Permit the inbound action — dispatch `kernel.send`.
    case allow
    /// Refuse it; the dispatcher replies `{error, reason:"unauthorized"}`.
    case deny(String)
}

/// The inbound FILTER — decides whether the peer may perform an inbound action.
public protocol IngressRule: Sendable {
    func authorize(_ action: AuthAction) -> AuthDecision
}

/// The outbound DECORATOR — the token this leg PRESENTS on its own outbound `call`s
/// (stamped on the frame envelope by `forward`). `nil` ⇒ present nothing.
public protocol EgressRule: Sendable {
    func credential() -> String?
}

/// Thrown when a rule spec names an unknown type or is malformed — fails the boot
/// loudly rather than silently mis-securing.
public struct AuthPolicyError: Error, CustomStringConvertible {
    public let message: String
    public init(_ message: String) { self.message = message }
    public var description: String { message }
}

/// Normalize a rule spec to `(type name, env var)`. Absent/null/empty ⇒ `(nil, nil)`.
/// String ⇒ `(name, nil)`. Object ⇒ `(type|policy, env|token_env)`.
func parseSpec(_ spec: JSON?) throws -> (name: String?, tokenEnv: String?) {
    guard let spec else { return (nil, nil) }
    switch spec {
    case .null:
        return (nil, nil)
    case .string(let s):
        return s.isEmpty ? (nil, nil) : (s, nil)
    case .object(let o):
        guard let name = o["type"]?.asString ?? o["policy"]?.asString else {
            throw AuthPolicyError("rule object missing 'type'")
        }
        return (name, o["env"]?.asString ?? o["token_env"]?.asString)
    default:
        throw AuthPolicyError("unsupported rule spec")
    }
}

/// The rule TYPE name for reflect — never surfaces the rule's config. Absent ⇒
/// `def` (`allow_all` for ingress, `silent` for egress).
public func ruleName(_ spec: JSON?, default def: String) -> String {
    switch spec {
    case .some(.string(let s)) where !s.isEmpty: return s
    case .some(.object(let o)): return o["type"]?.asString ?? o["policy"]?.asString ?? def
    default: return def
    }
}

/// Content-blind constant-time compare given equal length (length is not secret —
/// same posture as Python's `hmac.compare_digest`).
func constantTimeEquals(_ a: String, _ b: String) -> Bool {
    let x = Array(a.utf8)
    let y = Array(b.utf8)
    if x.count != y.count { return false }
    var diff: UInt8 = 0
    for i in 0..<x.count { diff |= x[i] ^ y[i] }
    return diff == 0
}

/// Resolve the leg's INGRESS rule: `ingress_rule` if present, else the legacy `auth`
/// shorthand. Absent ⇒ AllowAll (back-compat).
public func resolveIngress(ingressRule: JSON?, auth: JSON?) throws -> IngressRule {
    try IngressRules.resolve(ingressRule ?? auth)
}

/// Resolve the leg's EGRESS rule: `egress_rule` if present, else the legacy `auth`
/// shorthand (so `auth:"password"` presents the group token). Absent ⇒ Silent.
public func resolveEgress(egressRule: JSON?, auth: JSON?) throws -> EgressRule {
    try EgressRules.resolve(egressRule ?? auth)
}

File: src/lib/rust/crates/fantastic-io-bridge/src/authorizer.rs
//! Bridge authorization — base types + the two rule registries (`ingress`/`egress`).
//!
//! Two independent, TYPED rules govern a leg, mirrored on the wire and enforced on
//! the RECEIVER:
//!   - an INGRESS rule (the inbound FILTER) — [`ingress::resolve`] → `authorize`,
//!     consulted at the read-loop choke point before an inbound `call` dispatches.
//!   - an EGRESS rule (the outbound DECORATOR) — [`egress::resolve`] → `credential`,
//!     consulted by `forward` to stamp this leg's token on the frame ENVELOPE (never
//!     the dispatched payload — the target never sees it).
//!
//! Each rule is typed in the record (`{"type": <name>, "env": <var>}`) and resolved
//! BY NAME from a registry — the [`ingress`] and [`egress`] modules, each a folder
//! of one-rule-per-file plus a `mod.rs` importer that registers names. Add a rule =
//! drop a file + one registry arm; the engine never changes. The record carries
//! `ingress_rule`/`egress_rule` (symmetric), or the legacy `auth` shorthand (sets
//! both sides). Rules are TRANSITIONAL (inline plumbing), not invocational (agents).

use serde_json::Value;

pub mod egress;
pub mod ingress;

/// One inbound request the peer is asking this leg to perform locally.
pub struct Action<'a> {
    /// `"call"` (gated) | `"watch"` | `"unwatch"`.
    pub kind: &'a str,
    /// The local agent id the peer addressed.
    pub target: &'a str,
    /// `payload["type"]` — the verb requested (e.g. `"reflect"`).
    pub verb: &'a str,
    /// The `auth_token` the peer attached to this call on the frame ENVELOPE, if any.
    pub token: Option<&'a str>,
}

/// An ingress rule's verdict on an [`Action`].
pub enum Decision {
    /// Permit the inbound action — dispatch `kernel.send`.
    Allow,
    /// Refuse it; the read loop replies `{error, reason:"unauthorized"}`.
    Deny(String),
}

/// The inbound FILTER — decides whether the peer may perform an inbound action.
/// Object-safe so the gate stays rule-agnostic (`Arc<dyn IngressRule>`).
pub trait IngressRule: Send + Sync {
    /// Authorize one inbound action.
    fn authorize(&self, action: &Action) -> Decision;
}

/// The outbound DECORATOR — the token this leg PRESENTS on its own outbound `call`s
/// (stamped on the frame envelope by `forward`). `None` ⇒ present nothing.
pub trait EgressRule: Send + Sync {
    /// The credential to present, or `None`.
    fn credential(&self) -> Option<String>;
}

/// Default env var for the `password` rule when the record names none.
pub(crate) const DEFAULT_TOKEN_ENV: &str = "FANTASTIC_GROUP_TOKEN";

/// Normalize a rule spec `Value` to `(type_name, token_env)`. Absent/null/empty ⇒
/// `(None, None)`. String ⇒ `(name, None)`. Object ⇒ `(type|policy, env|token_env)`.
pub(crate) fn parse_spec(spec: Option<&Value>) -> Result<(Option<String>, Option<String>), String> {
    let Some(spec) = spec else {
        return Ok((None, None));
    };
    match spec {
        Value::Null => Ok((None, None)),
        Value::String(s) if s.is_empty() => Ok((None, None)),
        Value::String(s) => Ok((Some(s.clone()), None)),
        Value::Object(o) => {
            let name = o
                .get("type")
                .or_else(|| o.get("policy"))
                .and_then(Value::as_str)
                .ok_or_else(|| "rule object missing 'type'".to_string())?;
            let env = o
                .get("env")
                .or_else(|| o.get("token_env"))
                .and_then(Value::as_str)
                .map(str::to_string);
            Ok((Some(name.to_string()), env))
        }
        other => Err(format!("unsupported rule spec {other:?}")),
    }
}

/// The rule TYPE name for reflect — never surfaces the rule's config. Absent ⇒
/// `default` (`allow_all` for ingress, `silent` for egress).
pub fn rule_name(spec: Option<&Value>, default: &str) -> String {
    match spec {
        Some(Value::String(s)) if !s.is_empty() => s.clone(),
        Some(Value::Object(o)) => o
            .get("type")
            .or_else(|| o.get("policy"))
            .and_then(Value::as_str)
            .unwrap_or(default)
            .to_string(),
        _ => default.to_string(),
    }
}

/// Content-blind constant-time compare given equal length (length is not secret —
/// same posture as Python's `hmac.compare_digest`). Avoids a timing oracle without a
/// crypto dependency.
pub(crate) fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}
